### ISSUE TITLE
Create utility windows?

### ISSUE THREAD

**Issue by orange451** [glfw/glfw#1476]
Is it possible to create utility windows in GLFW?

i.e
![](https://code.makery.ch/blog/javafx-dialogs-official/minimal-decorations.png)

**Comment by CaffeineViking** [comment]
No, as far as I know there's no such thing in the GLFW API. You can disable maximization by using a Window [hint](https://www.glfw.org/docs/latest/window_guide.html#window_hints) such as `glfwWindowHint(GLFW_RESIZABLE, GLFW_FALSE)` before creating the window, but I don't think that is enough to solve your problem. It's probably not so easy to solve in a portable way, but by getting a native handle you can probably do it. After a bit of searching, here is how I would go about doing this on Windows:

1. Get the native window handle via `glfwGetWin32Window`, it's documented over [here](https://www.glfw.org/docs/latest/group__native.html) (and other platforms).
2. This will give you the `HWND`. The idea is to use that to do `windows.h`-related stuff (you'll probably want a couple of `#define` and `#ifdef` for `WIN32` so you can support more platforms later with more of these).
3. I'm not an Win32 expert, but it seems like [`EnableMenuItem`](https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-enablemenuitem) is what you're looking for. See [this here](https://devblogs.microsoft.com/oldnewthing/20100604-00/?p=13803) post.
4. Use that to disable the buttons you don't want. For the title bars, there's probably functions for that too.

Anyway, no guarantees ;-). But it seems like something like that ***should*** work. IDK if it's within GLFW's scope.

**Comment by orange451** [comment]
I sure hope it could be something that's supported by GLFW in the future. This isn't a windows-only style. It exists as well on mac:

![java](https://user-images.githubusercontent.com/5247778/57165194-652d5c80-6dc4-11e9-97dc-bd3b75365eb8.png)

**Comment by CaffeineViking** [comment]
Oh wow, MacOS was one of the platforms I didn't think supported that, my bad :-P. Still though, e.g. on the tiling Linux setup I've got, providing new window hints for that wouldn't do anything. Of course on an more usual Linux WM it should still be possible. There's even a [window property](https://specifications.freedesktop.org/wm-spec/1.3/ar01s05.html) you can pass into X11 to hint for that. I can see why it'd be useful, and it seems to be supported by most platforms (even if on some it might not do anything). So maybe like a additional window hint: `glfwWindowHint(GLFW_MINIMIZABLE, GLFW_FALSE)`? or some sort of window style thing?: `glfwWindowHint(GLFW_WINDOW_STYLE, GLFW_UTILITY_WINDOW)` or the like?

I guess the problem is where to draw the line, otherwise GLFW will start to look like a full blown GUI toolkit replacement in a few years (which it isn't). The thing you are proposing isn't outrageous though, as there is some window customizability in GLFW already. I'd love to hear the opinion of the core maintainers on this.

**Comment by orange451** [comment]
Also for clarification (just in case), the OK button was something I created in JavaFX. A Utility window is empty, like any other window.

**Comment by CaffeineViking** [comment]
Yeah I got that. It'd be a different story if you were proposing to add widgets :-P. That's why this is a "grey" zone right now. It makes sense to be able to disable minimization/maximization buttons and even change the size of the title bar. Again, the problem is where to draw the line on window customization I think. Will GLFW also need to support customizing stuff like making [frames thicker](https://docs.microsoft.com/en-us/windows/desktop/dwm/customframe#extending-the-client-frame) too? IMO that would be overkill. I actually agree with you on adding a portable way to enable/disable the minimization/maximization. I am a bit less enthusiastic on the big/small title bar thing though, as I feel going in that direction will quickly get out of hand.

**Comment by orange451** [comment]
Well my idea for using it is for small multi windows that will exist on top of the main window. If you’ve ever used Paint.net, I plan to do something similar with the utility windows.

https://upload.wikimedia.org/wikipedia/en/c/c9/Paint.net_screenshot.png

**Comment by CaffeineViking** [comment]
Paint.NET is freaking awesome :-) I've been using it since it's first release. Too bad there isn't any support for Linux though (the closest things I've found is Pinta). Yeah, I'd say the most reasonable way to solve this is by having a window hint to specify the "type" of the window. Adding individual knobs will very quickly become overwhelming I think. This can also be used to solve issue #293. More styles / types can be added if needed.

Something like `glfwWindowHint(GLFW_STYLE, GLFW_UTILITY)` and `glfwWindowHint(GLFW_STYLE, GLFW_TOOLBAR)`?

Seems like `UTILITY` and `TOOLBAR` have corresponding API hints in X11 [1], Win32 [2, 3] and AppKit [4]. They are called a bit differently, but they seem to do more or less the same thing [citation needed]. I guess here is also a place where we need to be careful, some of the window hints in those APIs are very specific and won't have the same functionality as in the other APIs. Some hints in [2, 3, 4] are also too fine grained to be useful.

1. [_NET_WM_WINDOW_TYPE (X11)](https://specifications.freedesktop.org/wm-spec/wm-spec-1.3.html#idm140130317606816)
2. [Window Styles (Win32)](https://docs.microsoft.com/en-us/windows/desktop/winmsg/window-styles)
3. [Extended Window Styles (Win32)](https://docs.microsoft.com/en-us/windows/desktop/winmsg/extended-window-styles)
4. [NSWindow.StyleMask (AppKit)](https://developer.apple.com/documentation/appkit/nswindow/stylemask)

**Comment by orange451** [comment]
You're much more knowledgable about this stuff than I am, haha. Thanks for pointing out all the technical stuff ^^

**Comment by wormyrocks** [comment]
Hey, I needed to do this for a work project and I have a cross-platform patch that will create utility windows on every platform.
My fork also does a handful of other things (including fixing what I believe to be a bug in macOS monitor enumeration) but the main one is adding an extra window hint GLFW_HIDE_FROM_TASKBAR - you can probably figure out what is relevant to that and what isn't.
https://github.com/glfw/glfw/compare/master...Looking-Glass:master

If there's interest in having this as a real feature I can clean up the patch and submit it from my personal Github account.

**Comment by orange451** [comment]
Any news on this? :) @wormyrocks Would be insanely useful.

**Comment by orange451** [comment]
@elmindreda Is this possible? It would benefit my application, but I don't have the ability to add it myself--I'm using LWJGL, so AFAIK I'm bound by their build of GLFW from the master branch here.

**Comment by orange451** [comment]
Any update on this? This feature would be quite useful for many application types! @elmindreda
If not, I may have to switch libraries.

**Comment by Osiris-Team** [comment]
Sup its 2022, need this for my project, any news?
